### PR TITLE
Fix render order of interop views

### LINKED ISSUE #1521: Use of Compose over (on top of) Swing/AWT Heavyweight component
### Problem
Very often in modern user interfaces, you need to place some UI elements over the main component of the application, like 'locate to my position' button over the Map component (in GoogleMaps), or 'play/pause' button over VideoPlayer. It's common practice. 
But there is NO **normal way** of using jb-compose components over Swing/AWT Heavyweight components like 3D Maps, video players, etc... When you try to use the normal method of layering (Box composable) with Heavyweight Swing/AWT components, you will never see your Compose overlay content, because the Heavyweight Swing/AWT component always redraws your Compose layer. 

### Temporary solution
After many failed attempts I found the way of using Compose over Heavyweight Swing/AWT component (in my case it's 3DMap - WoldWindGLCanvas component). The trick is to use Swing/Compose switching for hierarchy root and then use Box component overlaying with 1) Heavyweight Swing/AWT component wrapped in SwingPanel and 2) Compose content wrapped into SwingPanel and switched back to ComposePanel (Swing/Compose switching). 
So the hierarchy looks like this: 
```
PSEUDO-CODE

SwingPanel {
   factory = ComposePanel {
       Box {
            SwingPanel{
               factory = { YourHeavyweightSwingAWTComponent() } // <--- 3D Map in my case
            }
            SwingPanel {
              factory = ComposePanel {
                 YourComposableOverlay() // <--- Your overlay
              }
            }
       }
   }
}
```

You can find the working sample in [this repository](https://github.com/yaroslavkulinich/jb-compose-heavyweight-overlay) in the _master_ branch. And you can watch how it works on [this YouTube video](https://youtu.be/TyeJSLBlovw). I used beta5 version of jb-compose.

**But starting from version 1.0.0-beta6-dev455 of jb-compose the temporary solution is not working.** 
You see the overlay, but you can't do any actions with it (all mouse events not recognized).
You can find described NOT working behavior in [this repo](https://github.com/yaroslavkulinich/jb-compose-heavyweight-overlay/tree/broken) in _broken_ branch. And watch how it looks like on [this YouTube video](https://youtu.be/l530IUaTbL8). I used the rc3 version of jb-compose.

### Request
Dear jb-compose developers. @igordmn . 
Please add the ability to use Compose components over Heavyweight Swing/AWT components in the normal way with Box component or similar. It would be a huge step! It's a missing piece for many devs. 

If there is no possibility to make it the normal way, please, return the ability to use Compose over Heavyweight components like described in my temporary solution. Because after 1.0.0-beta6-dev455 it doesn't work. And in my personal case, I need to stay on beta5 or use ugly UI techniques without overlaying. 

Thank you. You do a great job!

### LINKED ISSUE #2926: Allow z-ordering with Swing interop
Interop with Swing means any Swing content embedded in a Compose panel will always be at the top (max z-order). Content should be composited with the right z-order instead.

### LINKED ISSUE #3848: UIKitView is displayed on top of all other components
**Describe the bug**
`UIKitView` is displayed on top of all other components

**Affected platforms**
- iOS

**Versions**
- Kotlin version*: 1.9.0
- Compose Multiplatform version*: 1.5.10-rc01
- OS version(s)* (required for Desktop and iOS issues): iOS 16.4
- OS architecture (x86 or arm64): x86

**To Reproduce**
[Here](https://gist.github.com/artsmvch/1f8eeb80c4b4f8b23bff0fee0d75f963) is a `Compose` screen for iOS.
The screen is a stack of two boxes with video players inside them. Video players are implemented via `UIKitView`.
But the video player from the bottom box is displayed on top of all other components (See attached screenshot).

**Expected behavior**
The first video is not visible at all, because it is placed under the second video and the blue background.

**Screenshots**

Screenshot 2023-10-22 at 23 16 09

**Additional context**
If I remove the top video, the bottom video will not be visible (correct behaviour).

### LINKED ISSUE #4004: Only the first UIKitView in a Box is displayed properly on iOS
**Describe the bug**
Multiple UIKitViews cannot be displayed in a Box on iOS. Only the first UIKitView is displayed properly.

**Affected platforms**
- iOS

**Versions**
- Kotlin version*: 1.9.21
- Compose Multiplatform version*: 1.6.0-dev1296
- OS version(s)* (required for Desktop and iOS issues): iOS 17.0
- OS architecture (x86 or arm64): arm64

**To Reproduce**

Common
```Kotlin
@Composable
expect fun DemoView()
```
iOS
```Kotlin
@OptIn(ExperimentalForeignApi::class)
@Composable
actual fun DemoView() {
    Box(modifier = Modifier.fillMaxSize().background(Color.Yellow)) {
        UIKitView(
            factory = {
                UIView().apply {
                    backgroundColor = UIColor.blueColor
                }
            },
            modifier = Modifier.size(100.dp),
        )
        UIKitView(
            factory = {
                UIView().apply {
                    backgroundColor = UIColor.redColor
                }
            },
            modifier = Modifier.size(100.dp).offset(50.dp, 50.dp),
        )
    }
}
```
Android
```Kotlin
@Composable
actual fun DemoView() {
    Box(modifier = Modifier.fillMaxSize().background(Color.Yellow)) {
        AndroidView(
            factory = { context ->
                View(context).apply {
                    setBackgroundColor(Color.Blue.toArgb())
                }
            },
            modifier = Modifier.size(100.dp)
        )
        AndroidView(
            factory = { context ->
                View(context).apply {
                    setBackgroundColor(Color.Red.toArgb())
                }
            },
            modifier = Modifier.size(100.dp).offset(50.dp, 50.dp)
        )
    }
}
```

**Expected behavior**
On iOS, only the blue square is displayed properly, with the other square appearing to be rendered behind it and without color. I expect a red square to be displayed overlapping a blue square, like on Android.

### PR DESCRIPTION
**Pull request by MatkovIvan**
## Proposed Changes

Currently on both Desktop and iOS interop views are added to the view hierarchy in order to add nodes to Compose. It works only if all intersecting interop views were added at the same time (frame). So it's basically last-added - above-displayed. This PR changes this behavior in a way that it will respect the order inside Compose like regular compose elements.

**It does NOT make any changes in the ability to display Compose content above interop view on Desktop**, this fix was made in #915

Main changes:
- Unify a way to work with interop on Desktop (`SwingPanel`) and iOS (`UIKitView`)
- `LocalInteropContainer` -> `LocalUIKitInteropContainer` on iOS
- `LocalLayerContainer` -> `LocalSwingInteropContainer` on Desktop
- Reduce copy-pasting by moving `OverlayLayout` and `EmptyLayout`
- Remove overriding `add` method on `ComposePanel` and `ComposeWindowPanel` - it was required to redirect interop, but now it's not required and it's better to avoid changing default AWT hierarchy behaviour
- Do not use `JLayeredPane`'s layers anymore - it brings a lot of transparency issues (faced with it on Windows too after unrelated change). Sorting via indexes is used instead
- Add `InteropOrder` page to mpp demo

### How it works

It utilizes `TraversableNode` to traverse the tree in the right order and calculate the index based on interop views count that placed before the current node in the hierarchy. All interop nodes are marked via `Modifier.trackSwingInterop`/`Modifier.trackUIKitInterop` modifier to filter them from the `LayoutNode`s tree.

## Testing

Test: run reproducers from the issues or look at "InteropOrder" page in mpp demo

Desktop | iOS
--- | ---
Screenshot 2024-02-27 at 12 51 06 | Simulator Screenshot - iPhone 15 Pro - 2024-02-27 at 12 49 50

### Desktop

Fixes https://github.com/JetBrains/compose-multiplatform/issues/2926
Fixes https://github.com/JetBrains/compose-multiplatform/issues/1521#issuecomment-1965240338

### iOS

Fixes https://github.com/JetBrains/compose-multiplatform/issues/4004
Fixes https://github.com/JetBrains/compose-multiplatform/issues/3848

### CODE REVIEW

**Review by elijah-semyonov — APPROVED** [review]

File: compose/mpp/demo/src/commonMain/kotlin/androidx/compose/mpp/demo/App.kt [hunk 1]
```diff
@@ -54,6 +54,7 @@ val MainScreen = Screen.Selection(
     Screen.Example("GraphicsLayerSettings") { GraphicsLayerSettings() },
     Screen.Example("Blending") { Blending() },
     Screen.Example("FontRasterization") { FontRasterization() },
+    Screen.Example("InteropOrder") { InteropOrder() },
```
Comment (igordmn) [inline compose/mpp/demo/src/commonMain/kotlin/androidx/compose/mpp/demo/App.kt:57]: This demo doesn't always work correctly (at least on Windows):

(reproducible 1 in 3 cases):

- Press Red - it is hidden, but the Green is over Blue:
image

or

- Press Blue - it is hidden, but the Red is over Green:
image

Comment (MatkovIvan) [reply compose/mpp/demo/src/commonMain/kotlin/androidx/compose/mpp/demo/App.kt:57]: Fixed by force revalidation of Swing container

File: compose/mpp/demo/src/commonMain/kotlin/androidx/compose/mpp/demo/InteropOrder.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.unit.dp
+
+@Composable
+fun InteropOrder() {
```
Comment (igordmn) [inline compose/mpp/demo/src/commonMain/kotlin/androidx/compose/mpp/demo/InteropOrder.kt:37]: I tried this strange example with `useInteropBlending = true` on Windows:
```
        Box {
            if (red) {
                Box(Modifier.size(150.dp).offset(0.dp, 0.dp).background(Color.Red))
                TestInteropView(Modifier.size(150.dp).offset(0.dp, 0.dp), Color.Red)
            }
            if (green) {
                TestInteropView(Modifier.size(150.dp).offset(75.dp, 75.dp), Color.Green)
            }
            if (blue) {
                Box(Modifier.size(150.dp).offset(150.dp, 150.dp).background(Color.Blue))
            }
        }
```

And sometimes it has a wrong ordering:
image


Sometimes correct:
image

Is it expected or is it a bug?

**Review by igordmn — COMMENTED** [review]
Cool!

With `interopBlending=true` it works with Compose in the middle:
```
        Box {
            if (red) {
                TestInteropView(Modifier.size(150.dp).offset(0.dp, 0.dp), Color.Red)
            }
            if (green) {
                Box(Modifier.size(150.dp).offset(75.dp, 75.dp).background(Color.Green))
            }
            if (blue) {
                TestInteropView(Modifier.size(150.dp).offset(150.dp, 150.dp), Color.Blue)
            }
        }
```

The code looks good to me.

File: compose/mpp/demo/src/commonMain/kotlin/androidx/compose/mpp/demo/InteropOrder.kt (hunk 2 above)
Comment (MatkovIvan) [reply compose/mpp/demo/src/commonMain/kotlin/androidx/compose/mpp/demo/InteropOrder.kt:37]: A bug. Fixed by adding validate/repaint after adding of interop (this PR) + Fix from #1147

**Comment by maxrave-dev** [comment]
The order of the view works, but I can not click on the Compose View, which overlays on top Swing Panel

**Pull request merged**